### ISSUE TITLE
Wrong RTP timestamps when Opus is used in mono mode

### ISSUE THREAD

**Issue by janh** [alfredh/baresip#753]
When Opus is used in mono mode, the difference between subsequent RTP timestamps is half what it actually should be (i.e. for ptime of 20ms it is 480 instead of 960).

The timestamp difference for a packet is calculated here from the sample count:
https://github.com/alfredh/baresip/blob/86a44bba978b7b5ff1cba035ee8c7a15c69e09c8/src/audio.c#L454

Dividing by `tx->ac->ch` instead of `tx->ac->pch` fixes the issue for Opus. This should also be right in the general case, as `sampc_rtp` is dependent on the actual number of channels and the RTP clock rate is supposed to be independent of the number of channels. But I haven't tested if it actually works for all other codecs (especially MPA).

**Comment by alfredh** [comment]
many thanks for the report. I will have a look in the next couple of days.

during a call you can try to press `A` and look at the stream duration:

```
--- Audio stream ---
 tx:   encode: opus 48000Hz/2ch ptime=20ms S16LE
       aubuf: wish_sz=3840 cur_sz=0 filling=0 (cur 0.00ms, max 600.00ms, or 0, ur 0)
       source: audiounit,default S16LE
       time = 10.400 sec
 rx:   decode: opus 48000Hz/2ch S16LE
       aubuf: wish_sz=3840 cur_sz=5020 filling=0 (cur 26.14ms, max 160.00ms, or 0, ur 2)
       player: audiounit,default S16LE
       n_discard:0
       time = 10.279 sec
 audio tx pipeline:   audiounit ---> opus
 audio rx pipeline:   audiounit <--- opus
 audio dir=sendrecv pt_enc=96
 local: 0.0.0.0:4052, remote: 212.79.111.155:19500/212.79.111.155:19500
 mnat: (none) (connected=no)
 menc: (none) (secure=no)
```

if everything is normal, the values of `time` should be roughly equal
and also relatively close to call duration (realtime).

**Comment by juha-h** [comment]
Jan Hoffmann writes:

> When Opus is used in mono mode, the difference between subsequent RTP
> timestamps is half what it actually should be (i.e. for ptime of 20ms
> it is 480 instead of 960).

Does this affect sound quality?

**Comment by janh** [comment]
> if everything is normal, the values of `time` should be roughly equal
> and also relatively close to call duration (realtime).

The `time` shown for tx direction is about half the time shown for rx direction.


> Does this affect sound quality?

Yes (at least with Asterisk on the other side), audio is garbled.

**Comment by alfredh** [comment]
hi @janh many thanks for the report and the suggested fix.
it was indeed a bug and as you correctly pointed out, `ch` should be used
instead of `pch`. in fact `pch` is not used in the RTP timestamp calculation,
only the SDP signaling.

```c
	/* Convert from audio samplerate to RTP clockrate */
	sampc_rtp = sampc * tx->ac->crate / tx->ac->srate;

	/* The RTP clock rate used for generating the RTP timestamp is
	 * independent of the number of channels and the encoding
	 * However, MPA support variable packet durations. Thus, MPA
	 * should update the ts according to its current internal state.
	 */
	frame_size = sampc_rtp / tx->ac->ch;

	tx->ts_ext += (uint32_t)frame_size;
```

if you look at this code, it first divides by `ac->srate` so it should be correct
to also divide by `ac->ch`.

opus is the only codec affected by this. MPA has a special way of passing
the timestamp delta from module to core.

I pushed a fix in commit 83899c8b0a2b9a02df008e3f8ffbc360c3ca4a8c

please, feel free to test it

**Comment by janh** [comment]
I can confirm the issue is fixed now.
